rust-symcrypt/src/chacha.rs

```diff
@@ -0,0 +1,108 @@
+use std::vec;
+use symcrypt_sys;
+use crate::errors::SymCryptError;
+
+pub fn chacha20_poly1305_encrypt(
+    key: &[u8],
+    nonce: &[u8],
+    auth_data: Option<&[u8]>,
+    src: &[u8],
+    tag: &mut [u8; 16] // this is an out parameter.
```

Comment:
Created as an out parameter to match GCM functions

Reply:
Since the tag is fixed length, you can always return in the Result as [u8; 16]

Reply:
I can do that, want to hear thoughts from @eholk.

Reply:
Whats the debate? It must be exactly 16 bytes.

Reply:
debate to return the tag or have it as an out parameter

Reply:
I assume this returns a vec with the encrypted version of `src`, right? What is the tag used for? (Speaking of which, doc comments would help make this a little clearer :) )

I'd probably steer you towards returning the tag rather than having it as an out parameter, especially since this returns a `Result`. That makes it harder to run into mistakes where, for example, you only cared about the `tag` and then you forgot to check whether the function succeeded.

It's a little tricky here because you're returning two values, but tuples work well for that (`Result<(Vec<u8>, [u8; 16]), SymCryptError>`) or even making a one off struct so you can name the fields and make it clearer what each one means.

Reply:
the tag parameter in this case is an additional parameter that is outputed along with the encrypted result. The tag is used so that others can "check your work" and make sure that the encryption isnt tampered with in anyway. I can change to using the Tuple as that's what I had before.

Reply:
Changed return to `Result<(Vec<u8>, [u8; 16]), SymCryptError>`